Use sof sdw as default sdw machine

bardliao opened the pull request:
Use sof_sdw as default Intel SoundWire machine driver when there is no link adr matches in the acpi match table.

bardliao committed:
ASoC: Intel: export sof_sdw_get_tplg_files

The sof_sdw_get_tplg_files function is a callback of snd_soc_acpi_mach.
Export it to allow other modules to use it.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
ASoC: soc_sdw_utils: add name_prefix to asoc_sdw_codec_info struct

Currently, the codec name_prefix of Intel SoundWire machine driver is
from the ACPI match table. We can have it in the asoc_sdw_codec_info
struct as a default name_prefix of a codec if there is no corresponding
audio config found in the ACPI match table.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
ASoC: SOF: Intel: use sof_sdw as default SDW machine driver

If there is no SoundWire machine matches the existing acpi match table,
get the required machine data from the acpi table and construct the
link adrs and endpoints. Pass the data to the default Intel SoundWire
machine driver. And we don't need to add new item to the acpi match
table in common cases.
We will construct a dummy topology name. The dummy topology is just
used to extract the platform name for function topology and should not
be used.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
ASoC: soc_sdw_utils: export asoc_sdw_get_dai_type

asoc_sdw_get_dai_type() is quite useful to convert SDCA function types
to SDW DAI types. It can be used by other drivers.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

lgirdwood commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1117,13 +1119,127 @@ static void hda_generic_machine_select(struct snd_sof_dev *sdev,
+		return NULL;
+
+	for (i = 0; i < asoc_sdw_get_codec_info_list_count(); i++) {
+		if (sdw_device->id.part_id == codec_info_list[i].part_id) {
```
if (!) continue would help with indentation.

bardliao replied:
updated

lgirdwood commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1117,13 +1119,127 @@ static void hda_generic_machine_select(struct snd_sof_dev *sdev,
+	if (!adr_dev)
+		return NULL;
+
+	for (i = 0; i < asoc_sdw_get_codec_info_list_count(); i++) {
```
loop could do with some inline comments.

lgirdwood commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1117,13 +1119,127 @@ static void hda_generic_machine_select(struct snd_sof_dev *sdev,
+	int i, j;
+
+	link->mask = BIT(sdw_device->bus->link_id);
+	if (!index)
+		adr_dev = devm_kzalloc(dev, sizeof(*adr_dev), GFP_KERNEL);
+	else
+		adr_dev = devm_krealloc(dev, (struct snd_soc_acpi_adr_device *)link->adr_d,
+					(index + 1) * sizeof(*adr_dev), GFP_KERNEL);
```
can you comment why we allocate size index + 1.

lgirdwood commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1117,13 +1119,127 @@ static void hda_generic_machine_select(struct snd_sof_dev *sdev,
+			((u64)(sdw_device->id.sdw_version & 0xF) << 44) |
+			((u64)(sdw_device->bus->link_id & 0xF) << 48);
+
+	if (!strcmp(adr_dev[index].name_prefix, "AMP")) {
```
Can we comment why AMP is special here.

lgirdwood commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1117,13 +1119,127 @@ static void hda_generic_machine_select(struct snd_sof_dev *sdev,
+		(*amp_index)++;
+	}
+
+	if (!strcmp(adr_dev[index].name_prefix, "Left")) {
```
ditto, best to comment the special cases.

lgirdwood commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1201,7 +1317,39 @@ static struct snd_soc_acpi_mach *hda_sdw_machine_select(struct snd_sof_dev *sdev
+
+	link_mask = 0;
+	link_index = 0;
+	for (i = 0; i < peripherals->num_peripherals; i++) {
```
what are we doing here, best to comment it.

lgirdwood reviewed (commented):
Only minor questions. @charleskeepax good for you ?

charleskeepax commented on sound/soc/amd/acp/amd-acp63-acpi-match.c:
```diff
@@ -9,21 +9,21 @@
 #include "../mach-config.h"
 
-static const struct snd_soc_acpi_endpoint single_endpoint = {
+static struct snd_soc_acpi_endpoint single_endpoint = {
```
Do we actually need to remove the const on all of these hard coded entries? At least from the description where we are only creating a match if one doesn't exist that doesn't imply we would be changing an existing hard coded entry?

bardliao replied:
The problem is that I don't know how to convert a const struct to a non-const struct. For example, I will get below compile error if I don't remove the const of cs42l43_endpoints and try to force cast the type when set the .endpoints.
```
sound/soc/intel/common/soc-acpi-intel-arl-match.c:200:38: error: conversion to non-scalar type requested
  200 |                 .endpoints = (struct snd_soc_acpi_endpoint)cs42l43_endpoints,
      |                                      ^~~~~~~~~~~~~~~~~~~~~
```
It would be great if you could provide me some advice. I am more than happy if we don't need to remove the const on all of those hard coded entries.

charleskeepax commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1201,7 +1333,44 @@ static struct snd_soc_acpi_mach *hda_sdw_machine_select(struct snd_sof_dev *sdev
+
+	links = devm_kcalloc(sdev->dev, peripherals->num_peripherals, sizeof(*links), GFP_KERNEL);
+	if (!links)
+		return NULL;
```
Is this right? I feel like num_peripherals is the number of devices but links are the soundwire links in usage? Is the naming slightly confusing or are we allocating the right sizes here?

bardliao replied:
My idea was that as we don't know the link numbers at this moment. Allocating the number of devices can guarantee that we allocate enough space for the links array. But, I use a loop to get the number of the links now. Thanks for pointing it out.

bardliao commented:
SOFCI TEST

charleskeepax commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1117,14 +1119,144 @@ static void hda_generic_machine_select(struct snd_sof_dev *sdev,
+
+		adr_dev[index].endpoints =
+			devm_kzalloc(dev, codec_info_list[i].dai_num *
+				     sizeof(struct snd_soc_acpi_endpoint), GFP_KERNEL);
```
Is the problem here? If instead of assigning the endpoint pointer in adr_dev directly what if you made a local non-const pointer, then once you have filled in the endpoints you can then assign that to the const pointer in the adr_dev structure.

bardliao replied:
No, the problem is
```c
struct snd_soc_acpi_adr_device {
        const u64 adr;
        const u8 num_endpoints;
        const struct snd_soc_acpi_endpoint *endpoints;
        const char *name_prefix;
};
```
and
```c
struct snd_soc_acpi_link_adr {
        const u32 mask;
        const u32 num_adr;
        const struct snd_soc_acpi_adr_device *adr_d;
};
```
I will get `assignment of read-only member` error even if the adr_device and the link_adr themselves are not const.

charleskeepax replied:
Pretty sure the problem is there, this seems to build, although I haven't actually runtime tested it:

https://github.com/charleskeepax/linux/commit/ce250fbcac26e703c1bf9abbe780ac65c50c2804
https://github.com/charleskeepax/linux/commits/sof/fix/

And feel free squash that in where ever it goes.

bardliao replied:
Oh, yes. Just make required variables non const and keep `snd_soc_acpi_endpoint` and `snd_soc_acpi_adr_device` const. Thanks for the advice. I will revise my PR, thanks.

lgirdwood reviewed (commented):
LGTM

bardliao committed:
ASoC: soc-acpi: make some variables of acpi adr and link adr non-const

Currently, we use predefined snd_soc_acpi_link_adr tables to match the
link adr from ACPI table to select the machine driver and the topology.
However, with the mechanism, we need to create the snd_soc_acpi_link_adr
table for each audio config. The sof_sdw machine driver is used by
almost all Intel platforms with SOF and we can load required topology
file dynamically today. In other words, we can use sof_sdw machine
driver as the default machine driver for Intel SOF SoundWire codecs and
no need to create snd_soc_acpi_link_adr table for every new audio
configs.
To achieve it, we need to drop the const for some members and edit the
link adr and acpi adr data to match the data from the ACPI table.

Suggested-by: Charles Keepax <ckeepax@opensource.cirrus.com>
Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

charleskeepax commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1201,7 +1335,54 @@ static struct snd_soc_acpi_mach *hda_sdw_machine_select(struct snd_sof_dev *sdev
+		 */
+		if (i < peripherals->num_peripherals - 1 &&
+		    peripherals->array[i + 1]->bus->link_id != peripherals->array[i]->bus->link_id)
+			link_index++;
```
Rather than doing this which depends on the links being ordered, would it be reasonable to do something like:

link_index = link_mask & (BIT(peripherals->array[i]->bus->link_id) - 1);

Pretty sure that works.

bardliao replied:
But the link_index may exceed the array size if the link_id is not continued or it doesn't start from 0. Like link 2 and 3 are used.

charleskeepax replied:
Apologies it seems I forgot an hweight in my comment there:

	link_index = hweight(link_mask & (BIT(peripherals->array[i]->bus->link_id) - 1));

For example a system using links 1,3 would look like:

	link_index = hweight(link_mask & (BIT(link_id) - 1))) = hweight(0xA & (0x2 - 1)) = hweight(0xA & 0x1) = 0
	link_index = hweight(link_mask & (BIT(link_id) - 1))) = hweight(0xA & (0x8 - 1)) = hweight(0xA & 0x3) = 1

charleskeepax replied:
It basically counts the number of used links below the current link to get the link_index.

bardliao replied:
Wow, that is clever. Thanks for the suggestion.

charleskeepax commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1201,7 +1335,54 @@ static struct snd_soc_acpi_mach *hda_sdw_machine_select(struct snd_sof_dev *sdev
+				"link_index %d exceeds the link number %d link_mask #%x\n",
+				link_index, link_num, link_mask);
+			return NULL;
+		}
```
Then one can probably drop this check too, since link_mask is generated from the link_id's the only way you could hit this was if the peripheral list was modified in between.

charleskeepax commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1117,14 +1119,146 @@ static void hda_generic_machine_select(struct snd_sof_dev *sdev,
+		if (sdw_device->id.part_id != codec_info_list[i].part_id)
+			continue;
+
+		int amp_group_id = 1;
```
would be nicer to put this before the if, with the endpoints variable.

bardliao replied:
fixed

ujfalusi commented on include/sound/soc-acpi.h:
```diff
@@ -131,8 +131,8 @@ struct snd_soc_acpi_adr_device {
-	const u32 mask;
-	const u32 num_adr;
+	u32 mask;
+	u32 num_adr;
```
nitpick: the commit message can mention that to achieve what is described, we will drop the `const` for some members.

ujfalusi commented on sound/soc/intel/common/sof-function-topology-lib.c:
```diff
@@ -132,4 +132,4 @@ int sof_sdw_get_tplg_files(struct snd_soc_card *card, const struct snd_soc_acpi_
 
 	return tplg_num;
 }
```
commit  message last sentence needs rewording

ujfalusi commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1117,14 +1119,145 @@ static void hda_generic_machine_select(struct snd_sof_dev *sdev,
+		if (dai_type == dai_info->dais[i].dai_type)
+			return true;
+	}
+	dev_dbg(&sdw_device->dev, "Endpoint DAI type %d not found\n", dai_type);
```
this is not an error, right?

do we need to print this and would we want to print if it is found also?

bardliao replied:
No, it is not an error. I only print when the dai_type is not fond because that the dai_type will be found in most cases. And that is expected. Printing it makes users know exactly the reason why the endpoint doesn't show up.

ujfalusi commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1117,14 +1119,145 @@ static void hda_generic_machine_select(struct snd_sof_dev *sdev,
+			continue;
+
+		endpoints = devm_kzalloc(dev, codec_info_list[i].dai_num *
+					 sizeof(struct snd_soc_acpi_endpoint), GFP_KERNEL);
```
devm_kcalloc() ?

ujfalusi replied:
check allocation fail

bardliao replied:
fixed, thanks.

ujfalusi commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1117,14 +1119,145 @@ static void hda_generic_machine_select(struct snd_sof_dev *sdev,
+
+		/* Get name_prefix from codec_info_list[] */
+		adr_dev[index].name_prefix = devm_kasprintf(dev, GFP_KERNEL, "%s",
+							    codec_info_list[i].name_prefix);
```
check adr_dev[index].name_prefix == NULL ?

bardliao replied:
fixed, thanks.

ujfalusi commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1117,14 +1119,145 @@ static void hda_generic_machine_select(struct snd_sof_dev *sdev,
+	 */
+	if (!strcmp(adr_dev[index].name_prefix, "AMP")) {
+		adr_dev[index].name_prefix = devm_kasprintf(dev, GFP_KERNEL, "%s%d",
+							    adr_dev[index].name_prefix,
```
we might want to free up the old `adr_dev[index].name_prefix` after this?

ujfalusi commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1117,14 +1119,145 @@ static void hda_generic_machine_select(struct snd_sof_dev *sdev,
+			adr_dev[index].name_prefix = devm_kasprintf(dev, GFP_KERNEL, "%s%d",
+								    "AMP", *amp_index);
+			if (!adr_dev[index].name_prefix)
+				return NULL;
```
we seams to keep some overshadowed allocation around while the module is loaded, can this be optimized somehow, or do we care?

bardliao replied:
Move the common `name_pfefix` allocation after the special cases. So that we will allocate the `name_prefix` just once.

ujfalusi commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1201,7 +1334,54 @@ static struct snd_soc_acpi_mach *hda_sdw_machine_select(struct snd_sof_dev *sdev
+	}
+
+	mach->drv_name = "sof_sdw";
+	mach->sof_tplg_filename = "sof-sdw-generic.tplg";
```
we do have this topology file already, right?

bardliao replied:
Not yet, I will submit it soon.

bardliao replied:
https://github.com/thesofproject/sof/pull/10040

plbossart replied:
not clear what this generic topology is because it's only there to "avoid errors"....

bardliao replied:
Currently, we still test the existence of the topology file. That is the main reason of the tplg. But the topology still works for generic SDCA codecs.

ujfalusi commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1201,7 +1334,54 @@ static struct snd_soc_acpi_mach *hda_sdw_machine_select(struct snd_sof_dev *sdev
+	mach->mach_params.platform = dev_name(sdev->dev);
+	mach->get_function_tplg_files = sof_sdw_get_tplg_files;
+	dev_info(sdev->dev, "Use default SDW machine driver %s topology: %s\n",
+		 mach->drv_name, mach->sof_tplg_filename);
```
Do we print when we are not using the 'default SDW machine driver sof_sdw toopology: sof-sdw-generic.tplg'?
Do we need to print this in info level?

bardliao replied:
Removed

plbossart commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1117,14 +1119,153 @@ static void hda_generic_machine_select(struct snd_sof_dev *sdev,
+{
+	int i;
+
+	for (i = 0; i < sdw_device->sdca_data.num_functions; i++) {
```
clearly this function is only intended to work on platforms which do have information on SDCA. I think what you are trying to do is simplify integration for NEW platforms where SDCA is well supported on Windows, and I think you'd want to place limits on when this code is used, i.e. don't use it before LNL.

bardliao replied:
Thanks @plbossart for pointing this out. Add `if (chip->hw_ip_version < SOF_INTEL_ACE_2_0)` test now.

plbossart reviewed (commented):
the problem statement is not clear, you are referring to 'common' uses, but we have lots of specific platforms with little to no SDCA information. I think this code is well-intended but is too optimistic for legacy platforms between CNL and MTL, where most of the ACPI information for SoundWire is garbage at best - as measured by the quirks to modify the deviceID...

bardliao commented:
> the problem statement is not clear, you are referring to 'common' uses, but we have lots of specific platforms with little to no SDCA information. I think this code is well-intended but is too optimistic for legacy platforms between CNL and MTL, where most of the ACPI information for SoundWire is garbage at best - as measured by the quirks to modify the deviceID...

The intention is to give the machine driver selection function a second chance to use the default SDW machine driver when it can't find any matched machine driver from the ACPI match table. Then we don't need to add new match to the ACPI match table for new audio configs. I.e. We still need to add the match table in some special cases. But in general, we don't have to do so.

bardliao commented:
change: use unique_id + 1 as name_prefix surfix for Cirrus Logic AMPs. @charleskeepax @simontrimmer Please check if it meet your requirement, thanks.

simontrimmer commented:
> change: use unique_id + 1 as name_prefix surfix for Cirrus Logic AMPs. @charleskeepax @simontrimmer Please check if it meet your requirement, thanks.

@bardliao 
Just having a look and talking about it here, we have seen products that have UIDs that are not sequential which would lead to the name_prefix also being non sequential - doesn't ALSA UCM rely on consecutively numbered amplifiers?

rfvirgil commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1117,14 +1119,170 @@ static void hda_generic_machine_select(struct snd_sof_dev *sdev,
+			/* Use unique_id + 1 as name_prefix surfix for Cirrus Logic AMPs */
+			adr_dev[index].name_prefix = devm_kasprintf(dev, GFP_KERNEL, "%s%d",
+								    name_prefix,
+								    sdw_device->id.unique_id + 1);
```
It's not safe to use id.unique_id like this, because of the special behavior in the SoundWire core code if the ACPI only has one device of a type on a link the SoundWire core ignores the unique ID and stores a magic number in id.unique_id instead of the actual ID.
https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=linux-6.14.y&id=2e8c4ad1f04413a4a67ef10746a7566007d2ed55

bardliao replied:
Thanks, I will add a check.

bardliao replied:
Hmm, it seems the existing code doesn't store the actual addr. We need to find a way to fix it if we need to use unique id as part of the name_prefix.

rfvirgil commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1117,14 +1119,170 @@ static void hda_generic_machine_select(struct snd_sof_dev *sdev,
+
+	switch (sdw_device->id.mfg_id) {
+		case 0x01fa:
+			/* Use unique_id + 1 as name_prefix surfix for Cirrus Logic AMPs */
```
It's not guaranteed that manufacturers will assign consecutive unique IDs. In fact we know some devices that have non-consecutive unique ids, such as 1,2 and 5. These would give you non-consecutive prefixes (like AMP2, AMP3, AMP6) which will break UCM because it wants consecutive numbers.

bardliao replied:
Hmm, what does the name prefix look like? How do you deal with that situation currently?

rfvirgil replied:
Current laptops use the prefix to load the correct firmware files to each amp. So the prefix must be correct, but that is hardcoded into the match table.

But to prepare for your changes, we have sent two patches to our driver to use the physical SoundWire address to find the correct firmware in future, instead of the ALSA prefix:
https://lore.kernel.org/linux-sound/20250612121428.1667-1-rf@opensource.cirrus.com/T/#t

So:

-  for CURRENT laptops with match tables, the machine driver should continue using the match table to ensure that the correct prefix is added to each amp.
- For NEW laptops we will switch to the new firmware naming scheme so it should not matter which prefix is assigned to each amp. The only requirement is that they must be sequential (AMP1, AMP2, AMP3, ...) because of alsaucm requirement.

We will discuss this more in Cirrus.

bardliao replied:
Ok, cool. if there is a match table for the config, then it is for sure will be used. The PR just provides a fallback solution for those configs that without match table. I will make sure the prefix is sequential.

bardliao replied:
@rfvirgil The current code should meet the requirement. Could you test it on your end?

lgirdwood commented:
@bardliao any update on this ?

bardliao commented:
SOFCI TEST

stefdb commented:
Hi,

We tried testing this on a lnl laptop with CS35L56 amps and a CS42L43 codec, with the match removed.
The laptop was ASUS UX5406SA (SSID 10431E13), and we removed the lnl_cs42l43_l0_cs35l56_l23 and lnl_cs42l43_l0 matches.

We hit a few issues with this.
Firstly, because this laptop has a 2ch dmic connected to the Intel PCH, it appends -2ch to the tplg_filename, which fails during sof_test_topology_file(), since instead of looking for sof-sdw-generic.tplg it looks for sof-sdw-generic-2ch.tplg, which doesn't exist. 

Related to this, when it looks for the dmic function topology in sof_sdw_get_tplg_files(), it needs the platform name, which it obtains from the tplg_filename, which I think normally contains the platform - but in this case it does not, because the filename is sof-sdw-generic.tplg, so the obtained platform is "sdw" instead of lnl, so it cannot load the dmic function topologies.

We also observed that the IDs of the function topologies were all shifted, which we think is because of the feedback dai on the CS35L56 amp. This increased the ids for the dmic and hdmi topologies, which meant it couldn't find those files.

Thanks,
Stefan

simontrimmer commented:
Hi @bardliao,
I can run tests on the ux5406 if you'd like! Just pulling a couple of things out from Stefan's report

Do we need the built topology files to include the -Nch variants of sof-sdw-generic.tplg ? (Looking in there there are dai IDs which would make this annoying)

And the follow on about whether the functional topology files should have the platform in them or the generic sdw, do these contain generational related configuration ?

Cheers!
-Simon

simontrimmer commented:
Cool thanks @bardliao , I'll keep an eye out for things to try out

simontrimmer commented:
Hiya @bardliao,
I think you're after:

> [    3.514417] kernel: snd_soc_sdw_utils:asoc_sdw_init_dai_link: sof_sdw sof_sdw: create dai link SDW0-Playback-SimpleJack, id 0
> [    3.514462] kernel: snd_soc_sdw_utils:asoc_sdw_init_dai_link: sof_sdw sof_sdw: create dai link SDW0-Capture-SimpleJack, id 1
> [    3.514503] kernel: snd_soc_sdw_utils:asoc_sdw_init_dai_link: sof_sdw sof_sdw: create dai link SDW2-Playback-SmartAmp, id 2
> [    3.514544] kernel: snd_soc_sdw_utils:asoc_sdw_init_dai_link: sof_sdw sof_sdw: create dai link SDW2-Capture-SmartAmp, id 3
> [    3.514680] kernel: snd_soc_sdw_utils:asoc_sdw_init_dai_link: sof_sdw sof_sdw: create dai link dmic01, id 4
> [    3.514747] kernel: snd_soc_sdw_utils:asoc_sdw_init_dai_link: sof_sdw sof_sdw: create dai link dmic16k, id 5
> [    3.514800] kernel: snd_soc_sdw_utils:asoc_sdw_init_dai_link: sof_sdw sof_sdw: create dai link iDisp1, id 6
> [    3.514853] kernel: snd_soc_sdw_utils:asoc_sdw_init_dai_link: sof_sdw sof_sdw: create dai link iDisp2, id 7
> [    3.514900] kernel: snd_soc_sdw_utils:asoc_sdw_init_dai_link: sof_sdw sof_sdw: create dai link iDisp3, id 8

bardliao commented:
@simontrimmer If you add the match back, the dai links will be created according to the match's information which doesn't include the feedback endpoint. As a result, the ` SDW2-Capture-SmartAmp, id 3` will not be created. However, the default match will create all endpoints. And the DMIC and HDMI BE IDs depend on how many SDW BEs are created. That's why you the driver request different functional topology files.
https://github.com/thesofproject/sof/pull/10157 The PR add some missing function topologies.

Regarding the topology being used, yes, it will use the latest one with `get_function_tplg_files = sof_sdw_get_tplg_files;` because we set `get_function_tplg_files = sof_sdw_get_tplg_files;` to the default match, too. But I think we probably don't need the restriction as we just need a existing topology to pass the topology existence test.

bardliao commented:
@simontrimmer Totally agree that the default match should be generic. Not sure what is the issue you met, but the default match should be created according to the ACPI table and will not take the `links` from the existing match tables. It will just take the `sof_tplg_filename`. I have no idea how the existing matches will affect the behavior of the default match. Do you have an example that will cause the issue? I will try to reproduce it on my side.

bardliao committed:
ASoC: SOF: Don't print the monolithic topology name if function topology
may be used

It may confuse people if "Topology file:" is printed but the topology
file is not really used. We keep the print if the topology file is
missing thus user can know what topology is missed in the file system.
And increase the log level to info for the actual topology that is used.
So that user can get which topology is used.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao commented:
@simontrimmer If your concern is about the `[    3.129698] kernel: sof-audio-pci-intel-lnl 0000:00:1f.3:  Topology file:     intel/sof-ipc4-tplg/sof-lnl-rt713-l2-rt1320-l13.tplg` log, https://github.com/thesofproject/linux/pull/5417/commits/5876993a8ab934d9a0f24c5192bfd2a0f3b6d030 will change the log to `Topology file:     function topologies`. However, we don't know if the function topologies are really used at that moment.
We will print `[    3.379386] kernel: sof-audio-pci-intel-lnl 0000:00:1f.3: Using function topologies instead intel/sof-ipc4-tplg/sof-lnl-rt713-l2-rt1320-l13.tplg` when we know that the function topologies will be used. And print the used function topology names
```
[    3.379465] kernel: sof-audio-pci-intel-lnl 0000:00:1f.3: loading topology 0: intel/sof-ipc4-tplg/sof-sdca-jack-id0.tplg
[    3.383378] kernel: sof-audio-pci-intel-lnl 0000:00:1f.3: loading topology 1: intel/sof-ipc4-tplg/sof-sdca-2amp-id2.tplg
[    3.386937] kernel: sof-audio-pci-intel-lnl 0000:00:1f.3: loading topology 2: intel/sof-ipc4-tplg/sof-lnl-dmic-2ch-id4.tplg
[    3.389202] kernel: sof-audio-pci-intel-lnl 0000:00:1f.3: loading topology 3: intel/sof-ipc4-tplg/sof-hdmi-pcm5-id6.tplg
```

> The "Topology file:" being printed is the monolithic one that I understand won't really be used?

Correct, the function topology will be used if all required function topologies are present.

> Though, the "sof_sdw sof_sdw: dai_link" seem to be affected by what is in that monolithic topology?

No, you see different IDs from "sof_sdw sof_sdw: dai_link" is because the match is different. The default match adds all endpoints including the feedback endpoint that the codec/amp can support. But the existing match doesn't include the feedback endpoint.
The topology should match the dai links but will not affect the dai links.

> As is the "sof-hdmi-pcm5", which attempts to load different IDs (which it can't because I didn't have those files on the system at the time)

https://github.com/thesofproject/sof/pull/10157 for different IDs

Using the file name of an existing topology bring another benefit that we don't need to create new generic monolithic topology for every new platform. Or maybe we can remove the existence check at boot time then fail when loading topology if the required topology is missing. @ujfalusi What do you think?

simontrimmer commented:
@bardliao also, just thinking about it whilst writing my work journal - why are there match tables with the functional topology callbacks if they have a monolithic topology file that has to be present to pass a file check? In the future if they can be supported by the generic infrastructure I guess they could be removed entirely?

bardliao committed:
ASoC: sof-function-topology-lib: escalate the log when missing function
topoplogy

Function topology must be used if the ACPI match table is created based
on the reported ACPI table because the monolithic topology is not picked
for the audio configuration.
Escalate the log and ask the user to download the latest topologies.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

simontrimmer commented:
@bardliao Cool, I'll be able to test tomorrow but I see the message change, it'll be interesting to see what the failure case looks like now in the log.

on
> What in my mind is that new match tables will use a generic topology file which doesn't have to prefect match to the audio configuration. And the function topologies will be used. And yes, our goal is to use function topologies.

We have a load of PTL matches that we're considering upstreaming as we already have some matches that will

bardliao commented:
> So it didn't find a match table entry, said it was going to use functional topologies, couldn't find the file that I'd sabotaged and then It did fall back to using the monolithic topology that should only be used as a dummy to get the "lnl" platform and pass a file existence check.
> 
> This got us wondering.... if someone upgraded the kernel to a version that did this, but didn't upgrade SOF, what if the monolithic topology it chose didn't actually fail... and the answer is that it look like it succeeds, the codec and amp drivers appeared to initialise properly, we get a sound card that includes all the controls we'd expect, wireplumber has a Lunar Lake audio controller with a stereo sink that you can play to but there are no catastrophic error messages in the log. Wireplumber gets headset events and appears to switch between speakers and headphones.

I don't think it will break the existing devices if they upgrade the kernel without upgrading SOF. All the existing working devices have their own ACPI match table. So, the driver will find the specific match. And regarding the dummy topology, if an acpi mach with `.get_function_tplg_files` is selected, I would expect that the SoundWire part will work.
But yeah, I agree that the dummy topology should not be used.

> 
> Is there a way of not using the dummy monolithic topology? Replace the tlpg filename with the platform name string, bypass the filename check if it's only three characters and if the functional topology load fails bail out without using the dummy file?
> 

Yes, I will try to do it.

bardliao commented:
@simontrimmer PR updated.

simontrimmer commented:
Hiya @bardliao the well configured and error scenarios are a lot clearer now.

I'll, er, try to break it a bit more but I think that'll do the trick. @stefdb  ?

bardliao commented:
> Hi @bardliao, I just tried this out on a different laptop and found something a bit undesirable -
> 
> This product has a match table entry in the kernel and a monolithic topology file on the filesystem, but because of
> 
> > commit [d348b41](https://github.com/thesofproject/linux/commit/d348b4181cd15ed432c2ae7eb33ef1bb7dfd7527)
> > Author: Bard Liao [yung-chuan.liao@linux.intel.com](mailto:yung-chuan.liao@linux.intel.com)
> > Date:   Mon Apr 14 14:32:35 2025 +0800
> > ```
> > ASoC: Intel: soc-acpi-intel-arl-match: set get_function_tplg_files ops
> > 
> > The audio configs with multi-function SDCA codecs can use the
> > sof_sdw_get_tplg_files ops to get function topologies dynamically.
> > 
> > Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>
> > Reviewed-by: Liam Girdwood <liam.r.girdwood@intel.com>
> > Reviewed-by: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>
> > Reviewed-by: Péter Ujfalusi <peter.ujfalusi@linux.intel.com>
> > Link: https://patch.msgid.link/20250414063239.85200-8-yung-chuan.liao@linux.intel.com
> > Signed-off-by: Mark Brown <broonie@kernel.org>
> > ```
> 
> It'll try to use functional topologies first, which aren't on this filesystem
> 
> > [    3.864763] sof-audio-pci-intel-mtl 0000:00:1f.3:  Firmware file:     intel/sof-ipc4/arl/sof-arl.ri
> > [    3.864765] sof-audio-pci-intel-mtl 0000:00:1f.3:  Firmware lib path: intel/sof-ipc4-lib/arl
> > [    3.864766] sof-audio-pci-intel-mtl 0000:00:1f.3:  Topology file:     function topologies
> > ...
> > [    4.073069] sof_sdw sof_sdw: Failed to open topology file: intel/sof-ace-tplg/sof-sdca-jack-id0.tplg, you might need to
> > [    4.073074] sof_sdw sof_sdw: download it from https://github.com/thesofproject/sof-bin/
> > ...
> 
> but does seem to then fallback to the monolithic topology as the sound card does all hook up, but it doesn't print the topology that was used (intel/sof-ace-tplg/sof-arl-cs42l43-l0-cs35l56-l23.tplg)

Hmm, the topology name is printed with debug level.
https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/topology.c#L2528

> 
> Did we have to add functional topology callbacks to all the existing match entries? It makes sense to use it going forward but I'm wondering how much confusion this is going to cause

I added functional topology callbacks to the multi-function codecs match entries. Currently the sof_sdw_get_tplg_files callback only supports multi-function codecs.

simontrimmer commented:
I'm just weighing up continuing changing the level of debug messages vs removing the callback from matches (there are around 15 the moment in scope for the Cirrus codecs and amps).

The logic is ideally if there is a match it should use the monolithic topology specified in the match entry as this maintains the behaviour of products that have already been tested with customers and it's extremely unlikely that we'd remove topology files from sof-bin as that would break compatibility with older kernels. On the downside that does mean I have a load of matches and topologies to make and upstream for PTL products.

I'm extremely keen to not add matches for the future Cirrus devices and any that sneak in before end user products go to market we'd back out so that they only used functional topologies.

ranj063 commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1215,7 +1373,57 @@ static struct snd_soc_acpi_mach *hda_sdw_machine_select(struct snd_sof_dev *sdev
+	/* SDCA was not well supported in the BIOS before ACE2.0 */
+	if (chip->hw_ip_version < SOF_INTEL_ACE_2_0) {
+		return NULL;
+	}
```
unnecessary braces

ranj063 commented on sound/soc/sof/fw-file-profile.c:
```diff
@@ -73,6 +73,10 @@ static int sof_test_topology_file(struct device *dev,
 	if (!profile->tplg_path || !profile->tplg_name)
 		return 0;
```
typo in commit: "machine date"

ranj063 commented on sound/soc/sof/intel/hda.c:
```diff
@@ -1215,7 +1373,57 @@ static struct snd_soc_acpi_mach *hda_sdw_machine_select(struct snd_sof_dev *sdev
 
-	return NULL;
+	/* The function topology needs the tplg file name to get the platform name for PCH DMIC */
+	/* FIXME: find another way to get the platform name */
```
can we update struct sof_intel_dsp_desc with the platform name?

bardliao replied:
Good suggestion will update.

bardliao committed:
ASoC: SOF: add platform name into sof_intel_dsp_desc

The platform name will be used construct the topology name.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

simontrimmer commented:
Hi @bardliao ,
I think we're out of thoughts on this one, we're looking at a few things that came out of wider testing but they won't get in the way of this chain
-Simon
